### PR TITLE
added global cache

### LINKED ISSUE #64: Allow disabling cache
Allow toggling the cache on or off, globally or per run.

This requires saving the setting in a file somewhere. We should use boltdb to store all settings, including the token, as a key/value pair.

### LINKED ISSUE #84: Add token into config file
store token inside the `~/.config/ipinfo/config.json` file and allow to configure it using ```$ ipinfo config token=<tok>```

### PR DESCRIPTION
**Pull request by rm-Umar**
Closes #64 
Closes #84 

Added global configurations 

```bash
$ ./ipinfo config cache=enable
```

### CODE REVIEW

**Comment by rm-Umar** [comment]
I think using JSON is better for config stuff instead of creating a nested boltdb

**Comment by rm-Umar** [comment]
also, there's already a subcommand `cache` if you think that would be better for this purpose I can use that, then the command will look like 
```bash 
$ ./ipinfo cache enable 
```

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:12]: We discussed this and the UI model agreed was to do it like:

```bash
$ ipinfo config cache=enable
$ ipinfo config cache=disable
$ ipinfo config <key>=<value>
```

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:1]: As a side note, for the future you can push your PRs to this main `ipinfo/cli` repository instead of your fork.

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:48]: We should print the particular argument that was invalid.

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:30]: I think we can probably move the `getConfigDir` func in `utils_config` into this new file now, since it's all strongly related.

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [reply ipinfo/cmd_config.go:12]: Also, @rm-Umar can you make an issue to look into making it so that the IPinfo token we store currently in `~/.config/ipinfo/token` is also stored inside the new config file? That way it can be even manually configured in the future via:

```bash
$ ipinfo config token=<tok>

# but still allow login to accept token, etc. - it will just put into the config file.
$ ipinfo login
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:23]: So this logic is basically initializing the config to defaults if it doesn't exist right? I think we can make a function for atleast the defaults part separately, like:

```go
func NewConfig() Config {
	return Config{
		GlobalCache: true,
	}
}
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:42]: Should add like 1 newline between all these function-call + error-handling cases.

File: ipinfo/config.go
Comment (UmanShahzad) [reply ipinfo/config.go:23]: Also for these `SetConfig` you're currently not handling the returned error?

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:20]:
```suggestion
	}

```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:12]: Any reason it's called `GlobalCache` and not just `Cache`?

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:64]: Consistent naming scheme:

```suggestion
	var conf Config
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:69]: We can just return `conf` here and not remake a new object - we already made the object above and unmarshaled into it.

File: ipinfo/init.go
Comment (UmanShahzad) [inline ipinfo/init.go:6]: This also returns an error.

File: ipinfo/ipinfo_client.go
Comment (UmanShahzad) [inline ipinfo/ipinfo_client.go:22]: Think we can remove this extra newline, as the comment above it as well as the variable are strongly connected to the following logic.

File: ipinfo/main.go
Comment (UmanShahzad) [inline ipinfo/main.go:22]: It's probably good to just put this variable in `config.go`.

File: ipinfo/main.go
Comment (UmanShahzad) [inline ipinfo/main.go:71]: We should update the default help message, and anywhere else this should appear.

**Review by UmanShahzad — CHANGES_REQUESTED** [review]

**Comment by UmanShahzad** [comment]
@rm-Umar lemme know when this is ready; there's a couple of things we need to go back-n-forth on before we comfortably merge this.

**Comment by rm-Umar** [comment]
@UmanShahzad yup you can review it now

**Comment by UmanShahzad** [comment]
Will review this soon, apologies.

File: ipinfo/utils_session.go
Comment (UmanShahzad) [inline ipinfo/utils_session.go:3]: Pretty much all of these functions can be deleted now, and we should just use the config functions and variables directly instead.

**Review by UmanShahzad — CHANGES_REQUESTED** [review]
I know it's a lot of review but take it slowly/carefully.

File: ipinfo/cmd_logout.go
Comment (UmanShahzad) [inline ipinfo/cmd_logout.go:40]: This change also changes the semantics here; before, it would print "not logged in" if the token was never saved, but now it prints that only if the SaveConfig fails, which it usually will not. We should only print it if `gConfig.Token` was already an empty string before we forcefully made it into an empty string.

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:108]: The actual logic here is to read the token value from the old token file, so it would be better renamed to `ReadTokenFile`. The documentation should also be updated.

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:112]:
```suggestion
	}

```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:90]:
```suggestion

	if err := SaveConfig(gConfig); err != nil {
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:133]: This is not a huge deal but what it does is, if we had a variable named `err` in this function scope, doing `err :=` here will not override the value of that higher-scope `err`. This is not too important right here but better practice to use `:=` unless you explicitly intend on overriding the value of the higher-scoped `err` variable.

```suggestion
	if err := ioutil.WriteFile(configPath, jsonData, 0644); err != nil {
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:153]: Same change here as above for the same reason.

```suggestion
	if err := json.Unmarshal(data, &config); err != nil {
```

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:70]:
```suggestion

	// save config in bulk.
	if err := SaveConfig(gConfig); err != nil {
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:77]: One thing you may notice is that this case is only possible if the user has no `config.json` yet - if they already have it, that means we also already migrated it before. So this logic can actually be pulled up into the `if/else` clause above, in particular in the `if` branch, because that's the first time we generate `config.json` for a user, and the **only** time they may still have the token file.

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:107]:
```suggestion
// might be deleted in future release as `token` file is deprecated.
```

File: ipinfo/config.go
Comment (UmanShahzad) [inline ipinfo/config.go:45]:
```suggestion
// returns the path to the token file.
//
// might be deleted in future release as `token` file is deprecated.
```

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:49]: Note that we do `strings.ToLower(configStr[0])` below when doing comparisons, but not here, which could cause confusion sometimes. To simplify, just make a new variable `key` which is already a lowered version, and use only that throughout.

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:62]:
```suggestion
				return fmt.Errorf("err: invalid value %s; cache must be 'enabled' or 'disabled'", strings.ToLower(configStr[1]))
```

Although you may want to make a `val := strings.ToLower(configStr[1])` here first so it can be reused in the `switch` case too (but don't do this for the `token` case below, of course - value may be case-sensitive there).

File: ipinfo/cmd_config.go
Comment (UmanShahzad) [inline ipinfo/cmd_config.go:72]:
```suggestion
	}

```

**Review by UmanShahzad — CHANGES_REQUESTED** [review]

File: ipinfo/cmd_logout.go
Comment (UmanShahzad) [inline ipinfo/cmd_logout.go:37]: Note that this comment is now out-of-date; should update the comments here.

**Review by UmanShahzad — COMMENTED** [review]
This is looking good. I'll give it a test run and approve if all looks good!

**Comment by UmanShahzad** [comment]
@rm-Umar the last thing we're missing is completions for this.

```go
var completionsConfig = &complete.Command{
	Flags: map[string]complete.Predictor{
		"-h":     predict.Nothing,
		"--help": predict.Nothing,
	},
}
```

**Comment by UmanShahzad** [comment]
Let's add completions then we can merge - tested it so far otherwise and it looks good.

**Review by UmanShahzad — APPROVED** [review]

**Pull request merged**
